### PR TITLE
eth/tracers/js: goja tracer

### PR DESCRIPTION
**Pull request by s1na**
Latest bench:

```
name                                old time/op    new time/op    delta
Tracers/create-8                      9.34ms ± 3%    5.16ms ± 1%    -44.82%  (p=0.008 n=5+5)
Tracers/deepCalls-8                   63.1ms ± 1%    38.9ms ± 0%    -38.39%  (p=0.008 n=5+5)
Tracers/delegatecall-8                13.4ms ± 1%     7.7ms ± 1%    -42.80%  (p=0.008 n=5+5)
Tracers/innerCreateOogOuterThrow-8    11.8ms ± 0%     6.2ms ± 1%    -47.04%  (p=0.016 n=5+4)
Tracers/innerInstafail-8              9.35ms ± 8%    5.92ms ± 1%    -36.66%  (p=0.008 n=5+5)
Tracers/innerThrowOuterRevert-8       13.5ms ± 1%     7.2ms ± 2%    -47.16%  (p=0.008 n=5+5)
Tracers/oog-8                         10.4ms ± 1%     5.4ms ± 1%    -48.27%  (p=0.008 n=5+5)
Tracers/revert-8                      8.19ms ± 2%    4.88ms ± 0%    -40.39%  (p=0.008 n=5+5)
Tracers/revertReason-8                9.33ms ± 1%    5.16ms ± 0%    -44.72%  (p=0.008 n=5+5)
Tracers/selfdestruct-8                7.76ms ± 1%    5.21ms ± 1%    -32.91%  (p=0.016 n=5+4)
Tracers/simple-8                      11.0ms ± 6%     6.5ms ± 1%    -40.94%  (p=0.008 n=5+5)
Tracers/throw-8                       9.37ms ± 6%    5.16ms ± 0%    -44.93%  (p=0.008 n=5+5)

name                                old alloc/op   new alloc/op   delta
Tracers/create-8                       134kB ± 1%    2464kB ± 0%  +1738.41%  (p=0.008 n=5+5)
Tracers/deepCalls-8                   2.14MB ± 0%   10.61MB ± 0%   +396.56%  (p=0.008 n=5+5)
Tracers/delegatecall-8                 250kB ± 1%    3027kB ± 0%  +1111.10%  (p=0.008 n=5+5)
Tracers/innerCreateOogOuterThrow-8     333kB ± 0%    2701kB ± 0%   +711.24%  (p=0.008 n=5+5)
Tracers/innerInstafail-8              69.1kB ± 0%  2713.1kB ± 0%  +3828.84%  (p=0.008 n=5+5)
Tracers/innerThrowOuterRevert-8        284kB ± 1%    2831kB ± 0%   +896.80%  (p=0.008 n=5+5)
Tracers/oog-8                          154kB ± 1%    2437kB ± 0%  +1479.09%  (p=0.008 n=5+5)
Tracers/revert-8                      57.7kB ± 1%  2393.7kB ± 0%  +4044.88%  (p=0.008 n=5+5)
Tracers/revertReason-8                 109kB ± 2%    2420kB ± 0%  +2118.34%  (p=0.008 n=5+5)
Tracers/selfdestruct-8                39.2kB ± 1%  2527.6kB ± 0%  +6352.31%  (p=0.008 n=5+5)
Tracers/simple-8                       156kB ± 1%    2738kB ± 0%  +1658.64%  (p=0.008 n=5+5)
Tracers/throw-8                       95.8kB ± 1%  2416.2kB ± 0%  +2423.34%  (p=0.008 n=5+5)
```

### CODE REVIEW

**Comment by fjl** [comment]
Very interesting. I think some of the allocs are because of ToValue, but we need to measure it.

**Comment by fjl** [comment]
This looks very promising. I also looked into this a tiny bit today and found that another cause of slowdown is with BigInteger.js. Since goja is a bit slower than duktape at running actual JS computation, bigint operations take more time.

We could kind of work around this by reimplementing bigInteger.js with math/big, but not sure if it would really save much time/allocs in the end.

**Comment by holiman** [comment]
One thing that would be good to benchmark is: 

- Instantiation of a tracer, with (and without) preloading the `bigIntegerJS`. In duktape, that happens within `tracer.New`:

```
	tracer.vm.EvalString(bigIntegerJS)
	tracer.vm.PutGlobalString("bigInt")
```
On my machine, doing `tracer.New` takes `6-7 ms` _with_ bigint_ , but `0.8ms` without loading that. So it's a very non-negligible part of the time spent. I'm curious if it has the same (or larger) toll on goja. 

If it turns out that the actual execution of bigint js is a large part of the time, then other routes to investigate could be:
- Is it possible to evaluate only once, and somehow lock it in a global env and reuse that env? 
- Is it possible to clear and reuse one tracer instance (or perhaps a small pool of them)?

I suspect both of these options are easier with Goja, since we have more control over the code executing, being go and all. 

Also, https://github.com/olebedev/go-duktape is now archived and abandoned, which I think is fairly recent, which adds support for the case of switching away from it.

**Comment by holiman** [comment]
```
$ benchstat duktape-bigint goja-bigint 
name         old time/op  new time/op  delta
JSVM/all-6   7.00ms ±11%  4.00ms ± 2%  -42.86%  (p=0.000 n=10+9)
```
So on my machine, tracer instantiation (plus bigint preloading) goes from `7ms` to `4ms`.

**Comment by holiman** [comment]
Oh wow, bigint js really takes a toll on goja too: 
```
$ benchstat goja-bigint goja-nobigint 
name        old time/op  new time/op  delta
JSVM/all-6  4.00ms ± 2%  0.21ms ± 0%  -94.63%  (p=0.003 n=9+4)
```

File: eth/tracers/goja.go [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+	isIstanbul := env.ChainConfig().IsIstanbul(env.Context.BlockNumber)
+	intrinsicGas, err := core.IntrinsicGas(input, nil, create, isHomestead, isIstanbul)
+	if err != nil {
+		return
+	}
+	t.ctx["intrinsicGas"] = t.vm.ToValue(intrinsicGas)
+}
+
+// CaptureState implements the Tracer interface to trace a single step of VM execution.
+func (t *GojaTracer) CaptureState(env *vm.EVM, pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, rData []byte, depth int, err error) {
```
Comment (holiman) [inline eth/tracers/goja.go:362]: Don't forget to void calling this unless `step` is actally implemented

**Comment by dapplion** [comment]
@s1na 💕

**Comment by s1na** [comment]
So I enabled the JS tracer [tests](https://github.com/ethereum/go-ethereum/blob/ae8ff2661d917506ec29bd8f457884d483e2ce42/eth/tracers/js/tracer_test.go) for the goja tracer. These 2 cases fail:

```
=== RUN   TestGojaTracer                                                                                                                                                                                                                 tracer_test.go:135: testcase 0: expected return value to be '[{},{},{}]' got '[[],[],[]]', error to be '' got ''                                                                                                                                 code: {depths: [], step: function(log) { this.depths.push(log.memory.slice(-1,-2)); }, fault: function() {}, result: function() { return this.depths; }}                                                                 tracer_test.go:135: testcase 7: expected return value to be '' got '{"foo":{"foo":...}}}', error to be 'RangeError: json encode recursion limit    in server-side tracer function 'result'' got ''
                code: {step: function() {}, fault: function() {}, result: function() { var o={}; var x=o; for (var i=0; i<1000; i++){  o.foo={}; o=o.foo; } return x; }}
```

The second one is because of the limitations in duktape, so not really a failure. The first one is a bit weird because I expected a slice of memory to be a list `[]`, not an object `{}`. Then I checked and it seems in duktape in most cases where we push a buffer from go to js (i.e. memory, any address or hash with fixed length like `contract.GetCaller()`) it shows up on the js side as something like `{0: 1, 1: 7, 2: 5...}`, whereas the current goja implementation simply returns sth like `[1, 7, 5...]`. I find the current behaviour a bit weird, anyone knows why that's the case?

**Comment by s1na** [comment]
I've been tracing goerli transactions via both duktape and goja to compare the result for our various tracers. So far found a difference in prestate tracer (will update this list):

- https://gist.github.com/s1na/7c9bd636c3ac0722a26a81a281d335d8

**Comment by s1na** [comment]
I fixed one of the main differences we had between goja and duktape (see [comment](https://github.com/ethereum/go-ethereum/pull/23773#issuecomment-989903232)). Basically byte strings were being passed to JS as normal arrays and not buffers (or more concretely `Uint8Array`). I found no way to create these objects using Goja's go API https://github.com/dop251/goja/issues/379) so I'm doing a hack for now. And that's we have a JS function that takes in an array and returns a `Uint8Array` from it. We call that JS func `toBuf` from Go whenever we want to push a buffer to JS.

**Comment by s1na** [comment]
Note for reviewers: you'll see `panic`s dispersed in the code. They're all in native functions that are exposed to JS. Goja will raise those panics as a JS interrupt which can be caught on the JS side. If they're not they'll simply cause execution to stop and the error will be returned like any other JS error.

**Comment by s1na** [comment]
So for some reason the `prestateTracer` is slower in goja than in duktape. No idea why yet, if I had to guess it's inefficiencies in how built-in functions like `toContract`, etc. are implemented. For every other tracer goja is a bit faster:

```
# Duktape

       noopTracerLegacy  callTracerLegacy  prestateTracerLegacy  opcountTracer  unigramTracer
mean          68.165723         98.223947             96.123697      68.585523      79.273222
std           39.852465         86.367623             74.865044      40.145431      53.258374
min            8.000000          9.000000              9.000000       8.000000       8.000000
25%           36.000000         44.000000             46.750000      36.000000      40.000000
50%           61.000000         79.000000             81.000000      61.000000      70.000000
75%           93.000000        129.000000            126.000000      93.000000     107.000000
max          376.000000       1918.000000           1709.000000     423.000000     963.000000

# Goja

       noopTracerLegacy  callTracerLegacy  prestateTracerLegacy  opcountTracer  unigramTracer
mean          60.582691         77.329746            154.703670      61.275940      66.597191
std           37.869345         54.091886            146.952691      37.834356      40.654601
min            5.000000          6.000000              7.000000       5.000000       5.000000
25%           30.000000         38.000000             61.000000      30.000000      34.000000
50%           53.000000         68.000000            108.000000      54.000000      59.000000
75%           85.000000        105.000000            212.000000      86.000000      92.000000
max          256.000000       1366.000000           3543.000000     206.000000     446.000000
```

File: eth/tracers/js/goja.go [hunk 2]
```diff
@@ -0,0 +1,840 @@
...
+		return nil, err
+	}
+	return vm.ToValue(res), nil
+}
+
+func fromBuf(vm *goja.Runtime, buf goja.Value, allowString bool) ([]byte, error) {
+	switch exported := buf.Export().(type) {
+	case string:
+		if allowString {
+			return common.FromHex(exported), nil
```
Comment (fjl) [inline eth/tracers/js/goja.go:68]: We need to use a stricter parser than `FromHex` here. `hexutil.Decode` would be best. But let's do it in another PR.
